Generalizing exchange

stigrj opened the pull request:
Generalizing to complex orbitals and response.

The on-the-fly algorithms should now be correct for all combinations of:
- Real and Complex orbitals
- Ground state (D1) and linear response (D2)
- Static and dynamic response
- `K.apply(phi)` and `K.dagger(phi)`

but _only_ the ground state versions are correct with `mpi::world_size > 1`. The response versions require some extra functionality of the bank (I think, haven't looked carefully) so that several operators (`K^(0)` and `K^(1)`) can have orbitals in the bank at the same time (in a seamless manner, without knowing about each other).

NB: I have removed the old `setupInternal()` and renamed `setupInternal_bank() -> setupInternal()`. This function should be disregarded and replaced with your new algorithm.

stigrj committed:
Adapt to new mrcpp interface

stigrj committed:
Generalize exchange to complex orbitals D1/D2

Currently exchange is on-the-fly only, i.e. no
setupInternal() is used. Exchange now requires
bank whenever MPI > 1. The D2 exchange is currently
NOT correct when MPI > 1.

Tweak exchange_prec input:
Now a negative value means it will follow the
setup precision, which is the current precision
used in the SCF cycle.

stigrj committed:
Update MRCPP to v1.3.1

stigrj commented:
@gitpeterwind you should watch this repo as well :slightly_smiling_face:

gitpeterwind commented on src/qmoperators/two_electron/ExchangePotentialD1.cpp:
```diff
@@ -122,86 +120,81 @@ void ExchangePotentialD1::setupInternal(double prec) {
  * The exchange potential is (pre)computed among the orbitals that define the operator
  */
 void ExchangePotentialD1::setupInternal_bank(double prec) {
```
Note that the main loop in setupInternal_bank will be completely different in next version. (i.e. do not put too much effort to try to improve it)

gitpeterwind commented on src/qmoperators/two_electron/ExchangePotential.cpp:
```diff
@@ -259,10 +243,10 @@ void ExchangePotential::calc_i_Int_jk_P(double prec,
 
     int out = -1;
-    if (Ex.size() == Phi.size()) {
+    if (K.size() == Phi.size()) {
```
Normally "K" is used for the exchange operator(s). Here it is the result of the application of K. It will confuse a new reader.

stigrj committed:
Change helmholtz_prec default

This precision should in general be negative, which means
it will follow the precision in the current SCF cycle.

stigrj commented:
Do you think there's any reason to keep the non-bank algorithms? I guess the new things you are working on will require bank also for other things than exchange?

gitpeterwind commented:
There is a small advantage keeping the possibility to run the model on a single MPI. But I think it is not worth the extra complications. And yes, in the future more parts will need a bank (and even two!)
 So I am fine with deleting the non-bank parts.
The compatibility with the non-MPI version we should still keep, don't we ?

stigrj commented:
Yes, we should keep the `mpi::world_size == 1` compatibility, but I think (hopefully) this can be treated as a special case in the bank algorithms and not require two separate implementations.

stigrj commented:
I think this in principle should work for all cases:
- `D1` and `D2` operators
- `apply()` and `dagger()`
- `mpi::world_size() == 1` and `mpi::world_size() > 1`

but I need to do more testing tomorrow.

The general strategy is this:
1. `K.setup(prec)`: put orbitals into bank
2. `K.apply(phi)`: on-the-fly application while fetching from bank
3. `K.clear()`: clear orbitals from bank

stigrj commented:
I think I see the problem with the response calculation. We need to be able to do the following:
1. `K^0.setup()`, i.e. save unperturbed orbitals `phi_i` in bank
2. `for (each response cycle)`:
    - `K^1.setup()`, i.e. save perturbed orbitals `x_i` and `y_i` in bank
    - `K^0.apply(x_i)` and `K^0.apply(y_i)`, this will access `phi_i` from bank
    - `K^1.apply(phi_i)` and `K^1.dagger(phi_i)`, this will access `phi_i`, `x_i` and `y_i` from bank
    - `K^1.clear()` i.e. delete perturbed orbitals `x_i` and `y_i` in bank
3. `K^0.clear()`, i.e. delete unperturbed orbitals `phi_i` in bank

stigrj committed:
Update unit test references

stigrj committed:
Update integration test references

gitpeterwind reviewed (commented):
Excellent!
